File: pathfinder-api/src/main/java/de/cubbossa/pathapi/storage/cache/GroupCache.java
package de.cubbossa.pathapi.storage.cache;

import de.cubbossa.pathapi.group.NodeGroup;
import de.cubbossa.pathapi.misc.NamespacedKey;
import de.cubbossa.pathapi.misc.Range;

import java.util.Collection;
import java.util.Optional;
import java.util.UUID;

public interface GroupCache extends StorageCache<NodeGroup> {

  Optional<NodeGroup> getGroup(NamespacedKey key);

  Optional<Collection<NodeGroup>> getGroups(NamespacedKey modifier);

  CacheCollection<NamespacedKey, NodeGroup> getGroups(Collection<NamespacedKey> keys);

  Optional<Collection<NodeGroup>> getGroups();

  Optional<Collection<NodeGroup>> getGroups(UUID node);

  Optional<Collection<NodeGroup>> getGroups(Range range);

  void write(UUID node, Collection<NodeGroup> groups);

  void write(NamespacedKey modifier, Collection<NodeGroup> groups);

  void writeAll(Collection<NodeGroup> groups);

  void invalidate(UUID node);

  void invalidate(NamespacedKey modifier);
}

File: pathfinder-api/src/main/java/de/cubbossa/pathfinder/storage/cache/GroupCache.kt
package de.cubbossa.pathfinder.storage.cache

import de.cubbossa.pathfinder.group.NodeGroup
import de.cubbossa.pathfinder.misc.NamespacedKey
import de.cubbossa.pathfinder.misc.Range
import de.cubbossa.pathfinder.storage.cache.StorageCache.CacheCollection
import java.util.*

interface GroupCache : StorageCache<NodeGroup?> {
    val groups: Collection<NodeGroup>?

    fun getGroup(key: NamespacedKey): NodeGroup?

    fun getGroups(modifier: NamespacedKey): Collection<NodeGroup>?

    fun getGroups(keys: Collection<NamespacedKey>): CacheCollection<NamespacedKey, NodeGroup>

    fun getGroups(node: UUID): Collection<NodeGroup>?

    fun getGroups(range: Range): Collection<NodeGroup>?

    fun write(node: UUID, groups: Collection<NodeGroup>)

    fun write(modifier: NamespacedKey, groups: Collection<NodeGroup>)

    fun writeAll(groups: Collection<NodeGroup>)

    fun invalidate(node: UUID)

    fun invalidate(modifier: NamespacedKey)
}
